high-level.lisp

```diff
@@ -958,3 +958,44 @@ with upper left block with dimension P-by-Q. Returns the intermediate representa
         (inv (recurse (make-identity-matrix (matrix-rows m)) m (- power)))
         (recurse (make-identity-matrix (matrix-rows m)) m power))))
 
+(define-condition singular-matrix-error (error)
+  ((matrix :initarg :matrix :reader singular-matrix-error-matrix))
+  (:documentation "A singular matrix is used where non-singular expected.")
+  (:report (lambda (condition stream)
+             (format stream "Singular matrix used where non-singular expected~%~A"
+                     (singular-matrix-error-matrix condition)))))
+
+(defun solve (a b)
+  "Computes the solution column vector X to the system of linear
+equations A * X = B, where A is a square matrix and B is an NxM
+matrix, the columns of which are different right-hand sides to the
+above equation. 
+
+The three return values are
+  1. The solution X to each of the RHS B; 
+  2. The L and U factors from the A = P * L * U factorization; and
+  3. The pivot indices."
+  (check-type a matrix)
+  (check-type b matrix)
+  (assert (square-matrix-p a))
+  (with-foreign-objects ((n ':int32) (nrhs ':int32)
+                         (lda ':int32) (ldb ':int32)
+                         (info ':int32))
+    (setf (cffi:mem-ref n :int32) (matrix-rows a))
+    (setf (cffi:mem-ref nrhs :int32) (matrix-rows b))
+    (setf (cffi:mem-ref lda :int32) (mem-ref n :int32))
+    (setf (cffi:mem-ref ldb :int32) (mem-ref nrhs :int32))
+    (setf (cffi:mem-ref info :int32) 0)
+    (let ((a* (copy-matrix-storage (matrix-data a)))
+          (ipiv (make-int32-storage (cffi:mem-ref n :int32)))
+          (b* (copy-matrix-storage (matrix-data b))))
+      (magicl.cffi-types:with-array-pointers
+          ((a*-ref a*)
+           (ipiv-ref ipiv)
+           (b*-ref b*))
+        (lapack::%%zgesv n nrhs a*-ref lda ipiv-ref b*-ref ldb info)
+        (when (> (cffi:mem-ref info :int32) 0)
+          (error 'singular-matrix-error :matrix a))
+        (values (make-matrix :rows (matrix-rows b) :cols (matrix-cols b) :data b*)
+                (make-matrix :rows (cffi:mem-ref n :int32) :cols (cffi:mem-ref n :int32) :data a*)
```

Comment:
Suggestion: wrap the `values` form in a `let` where you declare `n` to be equal to `(matrix-rows a)` so as not to repeat the call to `cffi:mem-ref`.